Build: enable -Wdocumentation via isystem

Empact opened the pull request:
This converts some dependency includes to use -isystem, so that we can
enable additional compiler checks and warnings independent of
the conformance of those dependencies.

The included new check is -Wdocumentation, which was previously proposed under #13914. See #14103 for examples of issues this has identified.

Here's a significant list of other clang checks we could consider given this. Some additional checks would require additional -isystem conversions, e.g. LevelDB is not converted here:
-Wreserved-id-macro
-Wthread-safety-attributes #15556
-Wthread-safety-negative
-Wfloat-equal
-Wfloat-conversion
-Wsign-conversion
-Wstring-conversion
-Wcast-qual
-Wcast-align
-Wshorten-64-to-32
-Wconversion
-Wfloat-conversion
-Wdouble-promotion
~-Wshadow #15377~
-Wshadow-field-in-constructor
-Wunused-exception-parameter
-Wunreachable-code-break
-Wdocumentation-unknown-command
-Wgnu-zero-variadic-macro-arguments
-Wgnu-anonymous-struct
-Wglobal-constructors
-Wexit-time-destructors
-Wnon-virtual-dtor
-Wweak-vtables
-Wcomma
-Wmissing-variable-declarations
-Wundefined-func-template
-Wzero-as-null-pointer-constant #15112
-Winconsistent-missing-destructor-override
-Wswitch-enum
-Wcovered-switch-default
-Wmissing-noreturn
-Wextra-semi
-Wc++17-extensions
-Wpadded
-Wold-style-cast
-Wmissing-prototypes

Empact committed:
build: Optionally include dependency headers with -isystem

When configured with --enable-isystem.

Was necessary to split QT_INCLUDES into QT_INCLUDES and
QT_MOC_INCLUDES because moc does not understand -isystem, e.g.:

    Unknown options: isystem/usr/local/Cellar/qt/5.10.0_1/include/QtNetwork[...]

This does not convert all uses, but focuses on libraries which have triggered
warnings/errors when applying initial additional build checks: QT, Univalue, and Berkeley DB.
LevelDb requires additional measures as its code is compiled with the project warnings
via AM_CXXFLAGS.

Note -isystem should not be applied to /usr/include, see BITCOIN_SYSTEM_INCLUDE
for a helper to convert -I to -isystem with /usr/include excepted.

practicalswift commented:
Concept ACK

Thanks for working on this!

Empact commented:
For GCC, it has been available since 1994, the egcs days. https://github.com/gcc-mirror/gcc/commit/0330c077aec77a65faee5ba84eb84c721f92c223

For Clang, has been in the codebase since 2012, coinciding with 3.2 or 3.3, though it wasn't mentioned in the release notes I checked. https://github.com/llvm-mirror/clang/commit/a4ba28f3215b4ea70d0cf15026edbfb535e2f0d3

MSVC has had a version of this since 2017, but AFAIK this should not affect those builds. https://blogs.msdn.microsoft.com/vcblog/2017/12/13/broken-warnings-theory/

Empact commented:
Another option is `-cxx-isystem`, but I'm not sure there's a good reason to prefer it. Does build with all `-isystem` replaced with `-cxx-isystem` under current clang.

fanquake commented:
Concept ACK
Compiled on macOS, this doesn't have the problem with warnings that #13914 did.

laanwj commented:
> For GCC, it has been available since 1994, the egcs days. gcc-mirror/gcc@0330c07 
> For Clang, has been in the codebase since at least 2009, coinciding  with 2.7, though it wasn't mentioned in the release notes I checked. llvm-mirror/clang@33a33d8 

TIL! Thanks for checking. That's far back enough, given that the minimum releases supported according to `dependencies.md` are gcc 4.8 and clang 3.3.

> MSVC has had a version of this since 2017, but AFAIK this should not affect those builds. https://blogs.msdn.microsoft.com/vcblog/2017/12/13/broken-warnings-theory/

Right—MSVC is a different beast.

laanwj commented:
utACK—would be nice if @theuni could take a look at the build system changes.

Empact commented:
In `Makefile.leveldb.include`, I split `LEVELDB_CPPFLAGS_INT` from `LEVELDB_CPPFLAGS`, and only gave the latter `isystem` treatment. The former is used for building leveldb as a lib, in which case the includes are not system includes.

fanquake commented:
Looks like this has issues on two of the Travis linux builds (re-ran to make sure).

[build](https://travis-ci.org/bitcoin/bitcoin/jobs/474239060):
```
checking miniupnpc/upnpcommands.h usability... yes
checking miniupnpc/upnpcommands.h presence... yes
checking for miniupnpc/upnpcommands.h... yes
checking for upnpDiscover in -lminiupnpc... (cached) yes
checking miniupnpc/upnperrors.h usability... yes
checking miniupnpc/upnperrors.h presence... yes
checking for miniupnpc/upnperrors.h... yes
checking for upnpDiscover in -lminiupnpc... (cached) yes
../configure: 25050: ./configure.lineno: Syntax error: redirection unexpected
```

[build](https://travis-ci.org/bitcoin/bitcoin/jobs/474239063):
```
checking whether the Boost::System library is available... no
checking whether the Boost::Filesystem library is available... no
checking whether the Boost::Thread library is available... no
checking whether the Boost::Chrono library is available... no
checking whether the Boost::Unit_Test_Framework library is available... no
checking for dynamic linked boost test... no
checking for mismatched boost c++11 scoped enums... ok
configure: error: No working boost sleep implementation found.
```

Empact commented:
Thanks, the redirection issue was a dash incompatibility, and the boost issue was a case of `-isystem/usr/include`. I updated any include that could resolve to /usr/include to run through the new`BITCOIN_SYSTEM_INCLUDE` function.

This is ready for review.

```
configure:29481: checking whether the Boost::System library is available
configure:29506: clang++ -std=c++11 -c  -DDEBUG_LOCKORDER -DHAVE_BUILD_INFO -D__STDC_FORMAT_MACROS -isystem/usr/include conftest.cpp >&5
In file included from conftest.cpp:76:
In file included from /usr/include/boost/system/error_code.hpp:14:
In file included from /usr/include/boost/system/config.hpp:13:
In file included from /usr/include/boost/config.hpp:57:
In file included from /usr/include/boost/config/platform/linux.hpp:15:
/usr/bin/../lib/gcc/x86_64-linux-gnu/7.3.0/../../../../include/c++/7.3.0/cstdlib:75:15: fatal error: 'stdlib.h' file not found
#include_next <stdlib.h>
              ^~~~~~~~~~
1 error generated.
```
https://travis-ci.org/bitcoin/bitcoin/jobs/474239063#L6175

practicalswift commented:
utACK 72878aac48e87e17646fa1889c554e229d76c7bb

Empact commented:
Added Qt Dbus and Qt Test `isystem` treatment.

practicalswift commented:
utACK d4baebfc06bf375b64f30298b685b4d66f32cae8

theuni commented:
I'm really not a fan of disabling warnings for 3rd party headers, but I'm not completely opposed. I'm more afraid that this will be one of those changes that causes lots of subtle future grief due to edge-cases in libtool, automake, ccache, non-glibc builds, dependency generation (gcc's "-M*" options), etc.

Clang has the "system-header-prefix" option, which could be hooked up to depends with just a few lines. It doesn't seem to exist in GCC yet, but I'd be happy to add and upstream that feature.

Thoughts on that approach?

Empact commented:
Can you elaborate on what sorts of issues you expect? One option is to apply `isystem` as a build option (by running everything through `BITCOIN_SYSTEM_INCLUDE`). Then we could enable the stricter checks only if `isystem` is enabled, but keep the existing arrangement otherwise.

Empact commented:
Updated to put these changes behind `--enable-isystem`. Also added `-Werror=documentation` for when `--enable-werror` is also specified.

The idea is to eventually error on all warnings when both are specified, so we can have a travis run which strictly applies all warnings.

Empact commented on configure.ac:
```diff
@@ -507,10 +523,12 @@ case $host in
        AC_CHECK_PROG([PORT],port, port)
        if test x$PORT = xport; then
          dnl add default macports paths
-         CPPFLAGS="$CPPFLAGS -isystem /opt/local/include"
```
Note I changed this for consistency with the option.

Empact replied:
Opened #15175 for separate consideration.

Empact replied:
Now merged. 👍

theuni reviewed (commented):
@Empact I gave a few examples above about possible weird edge cases. Qt demonstrates another with ```include_next```. Non-system toolchains and sysroots containing ```-I``` are two others.

I'm really uneasy about this due to the diversity of compilation environments. But other than the things I pointed out, it looks ok to me on paper.

-0.

theuni commented on build-aux/m4/ax_boost_base.m4:
```diff
@@ -147,6 +147,7 @@ if test "x$want_boost" = "xyes"; then
     if test "$ac_boost_lib_path" != ""; then
        BOOST_LDFLAGS="-L$ac_boost_lib_path"
     fi
+    BITCOIN_SYSTEM_INCLUDE([BOOST_CPPFLAGS], [$BOOST_CPPFLAGS])
```
These changes should be done outside of ax_boost_base.m4. This is an upstream file.

Empact replied:
Fixed

theuni commented:
On second thought, as long as it's off by default, concept ACK. Thanks for the compromise :)

Empact commented on build-aux/m4/bitcoin_subdir_to_include.m4:
```diff
@@ -12,7 +12,8 @@ AC_DEFUN([BITCOIN_SUBDIR_TO_INCLUDE],[
     AC_MSG_RESULT([${newinclpath}])
     if test "x${newinclpath}" != "x"; then
-      eval "$1=\"\$$1\"' -I${newinclpath}'"
+      BITCOIN_SYSTEM_INCLUDE([newincl], ["-I${newinclpath}"])
```
I'm skeptical, I don't find docs or use of `AC_INCLUDE`, and `AC_REQUIRE` is about requiring that a function be called. I think this is good thanks to `AC_CONFIG_MACRO_DIR`.
https://www.gnu.org/software/autoconf/manual/autoconf-2.67/html_node/Prerequisite-Macros.html
https://www.gnu.org/software/automake/manual/html_node/Local-Macros.html

theuni replied:
Yes, you're right.

Empact commented on configure.ac:
```diff
@@ -325,6 +331,16 @@ if test "x$CXXFLAGS_overridden" = "xno"; then
   AX_CHECK_COMPILE_FLAG([-Wdeprecated-register],[NOWARN_CXXFLAGS="$NOWARN_CXXFLAGS -Wno-deprecated-register"],,[[$CXXFLAG_WERROR]])
   AX_CHECK_COMPILE_FLAG([-Wimplicit-fallthrough],[NOWARN_CXXFLAGS="$NOWARN_CXXFLAGS -Wno-implicit-fallthrough"],,[[$CXXFLAG_WERROR]])
+
+  ## Some warnings alert on violations in our dependencies. Only enable them when
```
Fixed

Empact commented on build-aux/m4/bitcoin_system_include.m4:
```diff
@@ -0,0 +1,19 @@
+    dnl Including /usr/include with -isystem is known to cause problems, e.g.
+    dnl breaking include_next due to, the directive changing inclusion order.
+    dnl https://bugreports.qt.io/browse/QTBUG-53367
+    $1=$(echo $2 | sed -e 's|-I|-isystem|g' -e 's|-isystem/usr/include|-I/usr/include|g')
```
Fixed the `-I` issue, and I'd prefer to leave the cross toolchain issue aside. The goal being CI runs, initially, then perhaps broad support in the future.

Empact commented on configure.ac:
```diff
@@ -325,6 +331,16 @@ if test "x$CXXFLAGS_overridden" = "xno"; then
+    if test "x$enable_werror" = "xyes" && test "x$CXXFLAG_WERROR" != "x"; then
+      AX_CHECK_COMPILE_FLAG([-Werror=documentation],[ERROR_CXXFLAGS="$ERROR_CXXFLAGS -Werror=documentation"],,[[$CXXFLAG_WERROR]])
+    else
+      AX_CHECK_COMPILE_FLAG([-Wdocumentation],[WARN_CXXFLAGS="$WARN_CXXFLAGS -Wdocumentation"],,[[$CXXFLAG_WERROR]])
```
I believe this is correct - it's the non-error, with-isystem case. So it activates -Wdocumentation as a non-error.

Empact commented:
Rebased for #15175

practicalswift commented:
utACK 1433ed22dad8940e1339d741039d5882196e37b0

What about enabling isystem for the macOS 10.10 Travis job? That way we'll guard against broken Doxygen comments getting merged.

```
diff --git a/.travis.yml b/.travis.yml
index 6181726fb..bdde7bbf7 100644
--- a/.travis.yml
+++ b/.travis.yml
@@ -132,9 +132,9 @@ jobs:
       name: 'macOS 10.10  [GOAL: deploy]'
       env: >-
         HOST=x86_64-apple-darwin14
         PACKAGES="cmake imagemagick libcap-dev librsvg2-bin libz-dev libbz2-dev libtiff-tools python-dev python3-setuptools-git"
         OSX_SDK=10.11
         RUN_UNIT_TESTS=false
         RUN_FUNCTIONAL_TESTS=false
         GOAL="deploy"
-        BITCOIN_CONFIG="--enable-gui --enable-reduce-exports --enable-werror"
+        BITCOIN_CONFIG="--enable-gui --enable-reduce-exports --enable-werror -enable-isystem"
```

Empact commented:
@theuni would you take another look? I believe I’ve addressed your feedback.

theuni commented on configure.ac:
```diff
@@ -332,6 +345,12 @@ if test "x$CXXFLAGS_overridden" = "xno"; then
+
+  ## Some warnings alert on violations in our dependencies. Only enable them when
+  ## isystem is in use to keep the build clean
+  if test "x$use_isystem" = "xyes" && test "x$werror_enabled" == "x"; then
```
You can just use ```test "x$enable_werror" = xyes``` here, no need for the new variable.

Empact replied:
Ah good point - I dropped `werror_enabled`, as it was extraneous. I do think testing that isystem is active & werror is inactive is appropriate here, as this is the warn-but-don't-error isystem case. Added some comments to further clarify and connect the cases.

theuni commented:
I believe you'll need to remove these two lines before this works correctly with depends builds:
https://github.com/bitcoin/bitcoin/blob/master/depends/config.site.in#L61

That should be OK, because all depends should be found anyway without requiring those paths.

theuni reviewed (commented):
Ok, thinking about it more, the cross-compile issue probably isn't that big of a deal, because those cross-compiler include paths are very unlikely to be tacked on by any part of the build process.

utACK after addressing the last round of comments.

theuni commented:
> I believe you'll need to remove these two lines before this works correctly with depends builds:
> https://github.com/bitcoin/bitcoin/blob/master/depends/config.site.in#L61
> 
> That should be OK, because all depends should be found anyway without requiring those paths.

Nevermind on this. The -isystem that inevitably gets added later will override the "-I" added here.

I'll remove these lines as part of another cleanup.

practicalswift commented:
utACK c0ee524b77eacd3e7a2c31be9a722c467d876293

Empact commented:
Rebased due to conflicts in #15377

Empact commented:
Rebased due to conflicts in #15112

Empact committed:
build: Enable -Wdocumentation if isystem is enabled

Or -Werror=documentation if isystem & werror are enabled.

Empact commented:
Fix documentation errors added in #14978 bd0dbe8763fc3029cf96531c9ccaba280b939445

Empact commented:
Fix more documentation errors, including some added in #14021.

practicalswift commented:
utACK b51f4e14137e45d8d52fbf0b925287dde45ee97b

Let's move forward with this one :-)

Empact commented:
@theuni could you take another look at this?

Empact commented:
Rebased, squashed the latter two commits.

Empact commented:
@dongcarl could you take a look at this, given your recent build investigations?

Sjors commented:
Concept ACK

Needs rebase now that openssl and protobuf are gone.

Sjors commented:
Consider cherry-picking the reverse of f327e82b0266b9ac455c8b82434528d670eee6a8

vasild commented on build-aux/m4/bitcoin_system_include.m4:
```diff
@@ -0,0 +1,19 @@
+dnl which is left unmodified to avoid order issues with /usr/include. See:
+dnl https://stackoverflow.com/questions/37218953/isystem-on-a-system-include-directory-causes-errors
+AC_DEFUN([BITCOIN_SYSTEM_INCLUDE],[
+  if test "x$use_isystem" = "xyes" && test "x$2" != "x"; then
```
`test "x$2" != "x"` seems unnecessary? If `$2` is empty, then the below `echo | sed ...` would still assign an empty string to `$1`?

vasild commented:
ACK 572db3ce8 I tested this on FreeBSD 12 with clang 9 and it works as expected: `-I` is replaced with `-isystem`, excellent!

This PR does 2 things - add a new `./configure` option (first commit 572db3ce8) and add a new warning flag (second commit 875f8e2c2).

NACK 875f8e2c2 I do not like that the new warning flag is enabled only if this and that configure options are turned on or off. 875f8e2c2 turns `--enable-isystem` into a multi-tool that replaces `-I` with `-isystem` and also enables some compiler flags, maybe, depending on another configure option (`--enable-werror`).

`--enable-isystem` should do one thing and do it well, and that is to just replace `-I` with `-isystem`, nothing more (as in the first commit 572db3ce8). There are a few very good reasons to have such an option:

* Warnings from external headers (boost, qt, etc) are already suppressed if they are installed in `/usr/include`. I guess this is the common case with Linux. So, we just add an option to achieve the same if external headers are installed in another location.
* The new option is `off` by default.
* It would allow us to reliably build our code with full `-Werror` (not a partial set of `-Werror=...`).

Empact commented:
> I do not like that the new warning flag is enabled only if this and that configure options are turned on or off.

`--enable-isystem` here has two modes: if `--enable-werror` is active, it will enable `-Werror=*`, if not, it will enable `-W*`. The difference being whether or not to error on a related failure.

fanquake commented:
Thanks for the contributions here, however I'm going to enable `-Wdocumentation` using a different approach.

MarcoFalke commented:
#21613
